pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,603 @@
+	stream discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient,
+	delta discoverypb.AggregatedDiscoveryService_DeltaAggregatedResourcesClient,
+) error {
+	log := c.log.With(logFeatureKey, "start-requests")
+	for _, typeUrl := range c.opts.BootstrapResources {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		var err error
+		if c.opts.UseSOTW {
+			req := &discoverypb.DiscoveryRequest{
+				Node:    c.node,
+				TypeUrl: typeUrl,
+			}
+			log.Debug("Stream send", "req", req)
+			err = stream.Send(req)
+		} else {
+			req := &discoverypb.DeltaDiscoveryRequest{
+				Node:    c.node,
+				TypeUrl: typeUrl,
+			}
+			log.Debug("Delta send", "req", req)
+			err = delta.Send(req)
+		}
+		if err != nil {
+			return fmt.Errorf("initial requests: stream send: %w", err)
+		}
+	}
+	return nil
+}
+
+// fetchResponses will pass messages from Recv() calls to queue until Context ctx is done.
+func (c *XDSClient) fetchResponses(ctx context.Context, errCh chan error,
+	stream discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient,
+	delta discoverypb.AggregatedDiscoveryService_DeltaAggregatedResourcesClient,
+) {
+	defer close(errCh)
+	log := c.log.With(logFeatureKey, "fetch-responses")
+	backoff := backoff.Exponential{
+		Min:        c.opts.MinBackoff,
+		Max:        c.opts.MaxBackoff,
+		ResetAfter: c.opts.BackoffReset,
+		Jitter:     true,
+		Name:       "xds-client-fetch-responses",
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		if c.opts.UseSOTW {
+			resp, err := stream.Recv()
+			if err != nil {
+				log.Error("Failed to receive message from stream", "error", err)
+				errCh <- err
+				backoff.Wait(ctx)
+				continue
+			}
+			select {
+			case <-ctx.Done():
+				return
+			case c.responseQueue <- resp:
+			}
+		} else {
+			resp, err := delta.Recv()
+			if err != nil {
+				log.Error("Failed to receive message from stream", "error", err)
+				errCh <- err
+				backoff.Wait(ctx)
+				continue
+			}
+			select {
+			case <-ctx.Done():
+				return
+			case c.deltaResponseQueue <- resp:
+			}
+		}
+	}
+}
+
+func (c *XDSClient) getAllResources(typeUrl string) (*xds.VersionedResources, error) {
+	return c.cache.GetResources(typeUrl, 0, "", nil)
+}
+
+// Observe adds resourceNames to watched resources of a given typeUrl.
+// It will be sent to a server asynchronyously.
+func (c *XDSClient) Observe(ctx context.Context, typeUrl string, resourceNames []string) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case c.observeQueue <- &observeRequest{typeUrl: typeUrl, resourceNames: resourceNames}:
+	}
+	return nil
+}
+
+func (c *XDSClient) prepareRequest(obsReq *observeRequest) (*discoverypb.DiscoveryRequest, error) {
+	curr, err := c.getAllResources(obsReq.typeUrl)
+	if err != nil {
+		return nil, fmt.Errorf("get resources: %w", err)
+	}
+	reqResourceNames := make(map[string]bool, len(curr.ResourceNames)+len(obsReq.resourceNames))
```

Comment:
```suggestion
	reqResourceNames := make(map[string]struct{}, len(curr.ResourceNames)+len(obsReq.resourceNames))
```
Or we could potentially use sets.Set[string]